[3.13] gh-117398: Statically Allocate the Datetime C-API (GH-119472)

ericsnowcurrently opened the pull request:
This is a backport of 3 commits that go together.

(cherry picked from commit a895756aec688c049a983199e2d7fb801d6502c8)
(cherry picked from commit b30d30c747df2bf9f1614df8e76db2ffdb24fcd8)
(cherry picked from commit a89fc26f4ac90156a1b108135f92dcf7698190f2)

<!-- gh-issue-number: gh-117398 -->
* Issue: gh-117398
<!-- /gh-issue-number -->

vstinner committed:
gh-117398: Move types to datetime state (#118606)

Move types to the datetime_state structure of the _datetime
extension.

ericsnowcurrently committed:
gh-117398: Statically Allocate the Datetime C-API (GH-119472)

neonene commented on Modules/_datetimemodule.c:
```diff
@@ -6933,8 +6983,7 @@ _datetime_exec(PyObject *module)
     }
-    PyObject *capsule = PyCapsule_New(capi, PyDateTime_CAPSULE_NAME,
-                                      datetime_destructor);
+    PyObject *capsule = PyCapsule_New(capi, PyDateTime_CAPSULE_NAME, NULL);
```
Two `PyMem_Free(capi);` operations under this line can be removed from main and 3.13?

encukou replied:
Thanks for the catch! A fix is in #119662

encukou committed:
gh-117398: gh-119655: datetime: Init static state once & don't free it (GH-119662)

- While datetime uses global state, only initialize it once.
- While `capi` is static, don't free it (thanks @neonene in https://github.com/python/cpython/pull/119641/files#r1616710048)
